### ISSUE TITLE
t.m.smtp.ESMTPClient doesn't honour requiresAuthentication if heloFallback is enabled

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7258]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)| @hawkowl reported|
|-|-|
|Trac ID|trac#7258|
|Type|defect|
|Created|2014-04-27 16:27:25Z|
|Branch|https://github.com/twisted/twisted/tree/esmtp-bugs-7258-2|

When implementing [#7257](https://github.com/twisted/twisted/issues/7257), I noticed that if heloFallback is enabled on the ESMTPClient, it will not honour the requiresAuthentication flag, as this is only checked when attempting to do TLS.

<details><summary>Searchable metadata</summary>

```
trac-id__7258 7258
type__defect defect
reporter__hawkowl hawkowl
priority__normal normal
milestone__ 
branch__branches_esmtp_bugs_7258_2 branches/esmtp-bugs-7258-2
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__ 
time__1398616045000000 1398616045000000
changetime__1421853117380550 1421853117380550
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_1"></a>|@hawkowl commented|
|-|-|

(In [42576]) Branching to esmtp-bugs-7258.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_2"></a>|@hawkowl removed owner|
|-|-|

Bugfixes from [#7257](https://github.com/twisted/twisted/issues/7257) ported, with more docstrings, and formatting enhancements while I was touching things.

Build forced, putting up for review.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban commented|
|-|-|

Many thanks for the changes! The documentation is much appreciated!

Please note that I am just starting to do reviews so this is not an authoritative review... just screening :)

---------

 twisted/mail/smtp.py

AFAIK private class members should also be documented so _tlsMode needs
documentation.... I got this as a feedback from glyph.

In esmtpEHLORequired, esmtpAUTHRequired, esmtpTLSRequired, esmtpAUTHMalformedChallenge is see that code and resp arguments are not used.
Do they need to be documented? Do they need to be defined at all?
This looks like a wasted of time.

Some for esmtpTLSFailed where resp is not used.

Does the registerAuthenticator or esmtpState_ehlo which looks like a procedure needs to document return value?

Maybe 'ESMTPSenderFactory' needs documentation for the new instance member: currentProtocol

--------

twisted/mail/test/test_smtp.py

Instead of 'Double check that HELO fallback works.' maybe the docstring should
state all preconditions in which fallback is activated. In a similar way
to other tests.

I see that these lines are repeated in all tests... maybe they can be refactored
into a method.... self.connectClient() or something. I don't think that it is
important for the test how the actual connections is done... but the test
should only care for the response.

```
    self.clientProtocol.makeConnection(transport)
    self.clientProtocol.dataReceived(b"220 localhost\r\n")
    transport.clear()
    self.clientProtocol.dataReceived(b"500 not a esmtp server\r\n")
```

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_4"></a>|@hawkowl commented|
|-|-|

(In [43678]) Branching to esmtp-bugs-7258-2.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_5"></a>|@hawkowl commented|
|-|-|

Thanks for the review, I never noticed it :(

I fixed up some docstrings, and documented that private property.

The extra comments are because this may be overridden by someone, and documenting what it takes explicitly is better than not.

The repeated lines in the tests help the readability, I think, so I'd prefer to keep them.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph commented|
|-|-|

No build results! Please force a build next time; it speeds up review latency by approximately 8 minutes :).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph commented|
|-|-|

`- has ssl` `+ has tls`: best line ever :).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph set owner to @hawkowl|
|-|-|

Thanks for fixing this, hawkowl.

Test results look good.

Some non-binding observations:

    1. Lots of great new docstrings in this branch!  Thanks for improving this situation somewhat :).
    2. I'm not sure I totally understand the point of changing to add this extra `_tlsMode` flag.  For example, did you know:
```
>>> p.transport.startTLS(ctx)
>>> ITLSTransport.providedBy(p.transport)
True
```
    It seems like the old check was adequate.
    3. `ESMTPDowngradeTestCase.test_...`: unless there's some protocol specification I'm unaware of, shouldn't it be "not _an_ esmtp server"?

Things that should be addressed:

    1. twisted.mail.test.test_smtp.test_requireHELOFallbackOperates: "Double check that HELO fallback works" is a bad test docstring in two ways: why are you ''double'' checking something?  That would be a duplicate test.  Also since it's the first test in the suite it seems like the other test would actually be the "double" check.  Secondly, never say "works" or "correctly" or "accurately" or anything like that; say specifically what you mean by "works".  "Is used", perhaps?
    2. twisted.mail.smtp.ESMTPSenderFactory.buildProtocol: "Build the protocol" is not a useful docstring for a method called `buildProtocol` already.  Describe some attributes of the protocol being built.  For example, it's far from obvious why `self.nEmails` is multipled by 2 and then added to 2 in the construction of the protocol.  You could also describe the manner in which it is built; for example: it is set on the factory unconditionally, that state-manipulation is interesting, and also new in this branch.  what happens if the same protocol is used twice? should it raise an exception if `currentProtocol` is already set?).

twistedchecker's happy, builders are happy, and the substantive comments here are pretty trivial things about docstrings, so please address those to add a bit more information and then go ahead and merge!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_9"></a>|@hawkowl commented|
|-|-|

(In [43712]) review comments (refs [#7258](https://github.com/twisted/twisted/issues/7258))

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_10"></a>|@hawkowl commented|
|-|-|

Thanks for the review, Glyph!

1. Thanks! Might as well doc it while I'm changing it.
1. The _tlsMode flag was always there, but the docstring has messed up the diffs a bit. Now it just sets it in new places where things happen. Maybe it ought to be a property that calls that? Not sure, but it seems okay for now.
1. Oh, yes. That's hawkie not knowing English ;)

And the other things:

1. Rewrote that to explain what it actually does :)
1. Rewrote this too, to try and explain it the best I can. The nEmails stuff is also in the regular SMTPSenderFactory, and is the log count, and is only used in debug? The fixed-length debug seems... icky. Shouldn't it go to an actual logger instead maybe? The setting on the factory is also to bring it in line with SMTPSenderFactory.

Thanks, I'll just rerun the buildbots to double check and merge.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_11"></a>|@hawkowl set status to `closed`|
|-|-|

(In [43714]) Merge esmtp-bugs-7258-2: t.m.smtp.ESMTPClient doesn't honour requiresAuthentication if heloFallback is enabled

Author: hawkowl
Reviewers: exarkun, adiroiban, glyph
Fixes: [#7258](https://github.com/twisted/twisted/issues/7258)
